hardware/kubernetes.go

```diff
@@ -0,0 +1,132 @@
+	if err != nil {
+		panic(err)
+	}
+	return client
+}
+
+// NewKubernetesClient creates a new KubernetesClient client instance. It launches a goroutine to perform synchronization
+// between the cluster and internal caches. Consumers can wait for the initial sync using WaitForCachesync().
+// See k8s.io/client-go/tools/clientcmd for constructing *rest.Config objects.
+func NewKubernetesClient(config *rest.Config, namespace string) (*KubernetesClient, error) {
+	opts := tink.GetServerOptions()
+	opts.Namespace = namespace
+
+	// Use a manager from the tink project so we can take advantage of the indexes and caching it configures.
+	// Once started, we don't really need any of the manager capabilities hence we don't store it in the
+	// KubernetesClient
+	manager, err := tink.NewManager(config, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	managerCtx, cancel := context.WithCancel(context.Background())
+	go func() {
+		if err := manager.Start(managerCtx); err != nil {
+			panic(err)
+		}
+	}()
+
+	client := NewKubernetesClientWithClient(manager.GetClient())
+	client.close = cancel
+	client.waitForCacheSync = manager.GetCache().WaitForCacheSync
+
+	return client, nil
+}
+
+// NewKubernetesClientWithClient creates a new KubernetesClient instance that uses client to find resources. The
+// Close() and WaitForCacheSync() methods of the returned client are noops.
+func NewKubernetesClientWithClient(client ListerClient) *KubernetesClient {
+	return &KubernetesClient{
+		client:           client,
+		close:            func() {},
+		waitForCacheSync: func(context.Context) bool { return true },
+	}
+}
+
+// Close stops synchronization with the cluster. Subsequent calls to all of k's methods have undefined behavior.
+func (k *KubernetesClient) Close() {
+	k.close()
+}
+
+// WaitForCacheSync waits for the internal client cache to synchronize.
+func (k *KubernetesClient) WaitForCacheSync(ctx context.Context) bool {
+	return k.waitForCacheSync(ctx)
+}
+
+// IsHealthy always returns true.
+func (k *KubernetesClient) IsHealthy(context.Context) bool {
+	return true
+}
+
+// ByIP retrieves a hardware resource associated with ip.
+func (k *KubernetesClient) ByIP(ctx context.Context, ip string) (Hardware, error) {
+	var hw tinkv1alpha1.HardwareList
+	err := k.client.List(ctx, &hw, crclient.MatchingFields{
+		tink.HardwareIPAddrIndex: ip,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if len(hw.Items) == 0 {
+		return nil, fmt.Errorf("no hardware with ip '%v'", ip)
+	}
+
+	if len(hw.Items) > 1 {
+		names := make([]string, len(hw.Items))
+		for i, item := range hw.Items {
+			names[i] = item.Name
+		}
+		return nil, fmt.Errorf("multiple hardware with ip '%v': [%v]", ip, strings.Join(names, ", "))
+	}
+
+	return &hardware{hw.Items[0]}, nil
+}
+
+// Watch is unimplemented.
+func (k *KubernetesClient) Watch(context.Context, string) (Watcher, error) {
+	return nil, errors.New("kubernetes client: watch is unimplemented")
```

Comment:
Why is watch not implemented here?

Reply:
`Watch` is only used by the gRPC service. I've been told its used by Equinix Metal but the metadata capability doesn't need it, I don't think, hence I've left it out of this PR to be implemented in a follow up.

The changes in this PR should get us off the ground with respect to machine provisioning.

Reply:
Resolving as I don't intend to address. Reopen if needed.